appsec/src/extension/logging.c

```diff
@@ -451,13 +454,26 @@ void dd_log_shutdown()
 {
     mlog(dd_log_debug, "Shutting down the file logging");
 
-#ifdef ZTS
-    tsrm_mutex_free(_mutex);
-    _mutex = NULL;
-#endif
-    if (_log_strategy == log_use_file && _mlog_fd != -1 &&
-        _mlog_fd > fileno(stderr)) {
-        int ret = close(_mlog_fd);
+    if (_log_strategy == log_use_file &&
+        atomic_load_explicit(&_initialized, memory_order_acquire) &&
+        _mlog_fd != -1 && _mlog_fd > fileno(stderr)) {
+        // no need for the mutex at this point
+        // all the zts workers should have been done
+        // we're guaranteed visibility of the write to _mlog_fd by _initialized
+        int fsync_ret;
+        int errno_fsync;
+        int ret;
+        do {
+            fsync_ret = fsync(_mlog_fd);
```

Comment:
Do you need to do fsync in the loop? Whether `close` was interrupted has no bearing on the outcome of `fsync`, although I guess it doesn't hurt either.

Reply:
yeah I guess i can move it out